File: scala/22.scala
import scala.collection.mutable.ListBuffer
object Solution {
    def generateParenthesis(n: Int): List[String] = {
      val l = ListBuffer[String]()
      def gen(n: Int, left: Int, res: String): Unit = {
        if (n == 0){
          l += res + List.fill(left)(')').mkString
        } else {
          gen(n-1, left + 1, res +'(')
          if (left != 0)
            gen(n, left - 1, res + ')')
        }
      }
      gen(n, 0, "")
      l.toList
    }
}

File: go/22.go
import "strings"

func generateParenthesis(n int) []string {
	return dfs(0, n, "")
}

func dfs(left, n int, cur string) []string {
	if n == 0 {
		return []string{cur + strings.Repeat(")", left)}
	}
	res := []string{}
	// push
	res = append(res, dfs(left+1, n-1, cur+"(")...)
	// pop
	if left > 0 {
		res = append(res, dfs(left-1, n, cur+")")...)
	}

	return res
}